core/App/ehsm_provider.h

```diff
@@ -324,4 +325,16 @@ apikey -- 32byte random number
 */
 ehsm_status_t Enroll(ehsm_data_t *appid, ehsm_data_t *apikey);
 
+/*
+Description:
+Generate HMAC(SHA-256) of payload with given apikey
+Input:
+cmk -- A symmetric cmk
+apikey -- 32byte random number (in plaintext, not encrypted)
+payload -- the payload, a string
```

Comment:
payload is based64 encoded string of the request data

Reply:
Currently the `payload` sent by `ehsm_key_server/function.js` is a pure string (without base64 encode). 

The `gen_hmac()` function and related functions are only called when verifying incoming request and verifying quote in NodeJS side.  It accepts an object and turn it into string by `params_sort_str()` function. Thus, the original `gen_hmac()` does not encode the string into base64 format before calculating HMAC.

Th implementation of `gen_hmac_safe()` uses the same logic to ensure the output of `gen_hac()` and `gen_hmac_safe()` is consistant. Should I encode it in the NodeJS side and decode it in the ehsm_core?

Reply:
yes, please do it in this way

Reply:
already fixed?

Reply:
Yes. Now the `ffi_generateHmac` function extracts `payload` from incoming request using `JSON2STRUCT` macro, which will auto decode with base64.

References:
- https://github.com/truc0/ehsm/blob/274-move-the-hmac-verification-into-enclave/core/App/ffi_operation.cpp#L60
- https://github.com/truc0/ehsm/blob/274-move-the-hmac-verification-into-enclave/core/App/ffi_operation.cpp#L1573